letsencrypt/acme/util.py

```diff
@@ -145,3 +80,121 @@ def __repr__(self):
         return '{0}({1})'.format(self.__class__.__name__, ', '.join(
             '{0}={1!r}'.format(slot, getattr(self, slot))
             for slot in self.__slots__))
+
+
+class ACMEObject(ImmutableMap):  # pylint: disable=too-few-public-methods
+    """ACME object."""
+    zope.interface.implements(interfaces.IJSONSerializable)
+    zope.interface.classImplements(interfaces.IJSONDeserializable)
+
+    def to_json(self):  # pragma: no cover
+        """Serialize to JSON."""
+        raise NotImplementedError()
+
+    @classmethod
+    def from_valid_json(cls, jobj):  # pragma: no cover
+        """Deserialize from valid JSON object."""
+        raise NotImplementedError()
+
+    @classmethod
+    def _decode_b64jose(cls, data, size=None, minimum=False):
+        try:
+            decoded = jose.b64decode(data)
+        except TypeError:
+            raise errors.ValidationError()
+
+        if size is not None and ((not minimum and len(decoded) != size)
+                                 or (minimum and len(decoded) < size)):
+            raise errors.ValidationError()
+
+        return decoded
+
+    @classmethod
+    def _encode_hex16(cls, data):
+        return binascii.hexlify(data)
+
+    @classmethod
+    def _decode_hex16(cls, data, size=None, minimum=False):
+        if size is not None and ((not minimum and len(data) != size * 2)
+                                 or (minimum and len(data) < size * 2)):
+            raise errors.ValidationError()
+        return binascii.unhexlify(data)
+
+    @classmethod
```

Comment:
These encode/decode csr/cert functions seem out of place to me.  Aren't there only a few ACMEObjects that need them?